Sources/NIOHTTP1/HTTPDecoder.swift

```diff
@@ -15,709 +15,573 @@
-            let index = self.calculateIndex(data: data, length: len)
-            assert(index >= 0)
-            self.slice = (index, len)
-        }
-    }
-}
-
-private protocol AnyHTTPDecoder: class {
-    var state: HTTPParserState { get set }
-    func popRequestMethod() -> HTTPMethod?
-}
-
-/// A `ChannelInboundHandler` used to decode HTTP requests. See the documentation
-/// on `HTTPDecoder` for more.
-///
-/// While the `HTTPRequestDecoder` does not currently have a specific ordering requirement in the
-/// `ChannelPipeline` (unlike `HTTPResponseDecoder`), it is possible that it will develop one. For
-/// that reason, applications should try to ensure that the `HTTPRequestDecoder` *later* in the
-/// `ChannelPipeline` than the `HTTPResponseEncoder`.
-///
-/// Rather than set this up manually, consider using `ChannelPipeline.addHTTPServerHandlers`.
-public final class HTTPRequestDecoder: HTTPDecoder<HTTPServerRequestPart>, RemovableChannelHandler {
-    public convenience init() {
-        self.init(leftOverBytesStrategy: .dropBytes)
-    }
-
-    /// Creates a new instance of `HttpRequestDecoder`.
-    ///
-    /// - parameters:
-    ///     - leftOverBytesStrategy: the strategy to use when removing the decoder from the pipeline and an upgrade was detected
-    public convenience init(leftOverBytesStrategy: RemoveAfterUpgradeStrategy) {
-        self.init(type: HTTPServerRequestPart.self, leftOverBytesStrategy: leftOverBytesStrategy)
-    }
+private enum HTTPDecodingState {
+    case beforeRequest
+    case beforeURL
+    case url
+    case headerName
+    case headerValue
+    case trailerName
+    case trailerValue
+    case headersComplete
 }
 
-/// A `ChannelInboundHandler` used to decode HTTP responses. See the documentation
-/// on `HTTPDecoder` for more.
-///
-/// The `HTTPResponseDecoder` must be placed later in the channel pipeline than the `HTTPRequestEncoder`,
-/// as it needs to see the outbound messages in order to keep track of what the HTTP request methods
-/// were for accurate decoding.
-///
-/// Rather than set this up manually, consider using `ChannelPipeline.addHTTPClientHandlers`.
-public final class HTTPResponseDecoder: HTTPDecoder<HTTPClientResponsePart>, ChannelOutboundHandler, RemovableChannelHandler {
-    public typealias OutboundIn = HTTPClientRequestPart
-    public typealias OutboundOut = HTTPClientRequestPart
-
-    /// A FIFO buffer used to store the HTTP request verbs we've seen, to ensure
-    /// we handle HTTP HEAD responses correctly.
-    ///
-    /// Because we have to handle pipelining, this is a FIFO buffer instead of a single
-    /// state variable. However, most users will never pipeline, so we initialize the buffer
-    /// to a base size of 1 to avoid allocating too much memory in the average case.
-    private var methods: CircularBuffer<HTTPMethod> = CircularBuffer(initialCapacity: 1)
+private class BetterHTTPParser {
+    var delegate: HTTPDecoderDelegate! = nil
```

Comment:
Why hold the protocol existential instead of making this class generic over the contained type?

Reply:
because the `@convention(c)` http_parser callbacks would then capture this generic param and they can't unfortunately...